compiler/x/codegen/FPTreeEvaluator.cpp

```diff
@@ -1328,131 +967,107 @@ TR::Register *OMR::X86::TreeEvaluator::fpConvertToLong(TR::Node *node, TR::Symbo
 
-      TR::TreeEvaluator::coerceFPOperandsToXMMRs(node, cg);
+   TR::Node        *child          = node->getFirstChild();
+   TR::Register    *sourceRegister = NULL;
+   TR::Register    *targetRegister = cg->allocateRegister(TR_GPR);
+   TR::LabelSymbol *startLabel     = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+   TR::LabelSymbol *endLabel       = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+   TR::LabelSymbol *exceptionLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
 
-      TR::Node        *child          = node->getFirstChild();
-      TR::Register    *sourceRegister = NULL;
-      TR::Register    *targetRegister = cg->allocateRegister(TR_GPR);
-      TR::LabelSymbol *startLabel     = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
-      TR::LabelSymbol *endLabel       = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
-      TR::LabelSymbol *exceptionLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+   sourceRegister = cg->evaluate(child);
+   generateRegRegInstruction(cvttOpCode, node, targetRegister, sourceRegister, cg);
 
-      sourceRegister = cg->evaluate(child);
-      if (sourceRegister->getKind() == TR_X87 && child->getReferenceCount() == 1)
-         {
-         TR_ASSERT(cg->comp()->target().is32Bit(), "assertion failure");
-         TR::MemoryReference  *tempMR = cg->machine()->getDummyLocalMR(TR::Float);
-         generateFPMemRegInstruction(TR::InstOpCode::FSTMemReg, node, tempMR, sourceRegister, cg);
-         generateRegMemInstruction(TR::InstOpCode::CVTTSS2SIReg4Mem,
-                                   node,
-                                   targetRegister,
-                                   generateX86MemoryReference(*tempMR, 0, cg), cg);
-         }
-      else
-         {
-         generateRegRegInstruction(cvttOpCode, node, targetRegister, sourceRegister, cg);
-         }
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
 
-      startLabel->setStartInternalControlFlow();
-      endLabel->setEndInternalControlFlow();
+   generateLabelInstruction(TR::InstOpCode::label, node, startLabel, cg);
 
-      generateLabelInstruction(TR::InstOpCode::label, node, startLabel, cg);
+   if (longTarget)
+      {
+      TR_ASSERT_FATAL(cg->comp()->target().is64Bit(), "We should only get here on AMD64");
+      // We can't compare with 0x8000000000000000.
+      // Instead, rotate left 1 bit and compare with 0x0000000000000001.
+      generateRegInstruction(TR::InstOpCode::ROL8Reg1, node, targetRegister, cg);
+      generateRegImmInstruction(TR::InstOpCode::CMP8RegImms, node, targetRegister, 1, cg);
+      generateLabelInstruction(TR::InstOpCode::JE4, node, exceptionLabel, cg);
+      }
+   else
+      {
+      generateRegImmInstruction(TR::InstOpCode::CMP4RegImm4, node, targetRegister, INT_MIN, cg);
+      generateLabelInstruction(TR::InstOpCode::JE4, node, exceptionLabel, cg);
```

Comment:
Put this JE4 (and the one in the if branch) after the if/else paths to reduce code footprint.